### ISSUE TITLE
graceful shut down

### ISSUE THREAD

**Issue by chpio** [jcrugzz/changes-stream#8]
hi,

what's the best practice for a graceful shut down?
im piping it to other stream handlers

**Comment by jcrugzz** [comment]
@chpio calling `stream.destroy()` should be graceful and end the stream (propagating down the pipe chain). Let me know if it does not seem to be so.

**Comment by mmalecki** [comment]
I was just looking at a related issue. It doesn't propagate on my node because we're calling `removeAllListeners` right after calling `push(null)`. I have a fixes for that which I'll PR shortly.

**Comment by jcrugzz** [comment]
@mmalecki good call. `removeListeners` in the nextTick should do the trick?

**Comment by jcrugzz** [comment]
@chpio let us know if its fixed in `2.0.0`

**Comment by darobin** [comment]
This could be a stupid question, but are the two calls to `removeAllListeners()` even needed? The documentation warns against those and indeed they create a problem.

We are using this module in order to pipe an event stream back over the Web through Express. It works great, except when the client terminates the connection (which can be as simple as closing the current browser tab). When that happens, the socket hangs up, triggering an `error` event on the `http` request object. There are no listeners left to handle it and so the `error` event crashes the process:

```
Error: socket hang up
    at createHangUpError (_http_client.js:250:15)
    at TLSSocket.socketCloseListener (_http_client.js:282:23)
    at emitOne (events.js:95:20)
    at TLSSocket.emit (events.js:182:7)
    at TCP._onclose (net.js:475:12)
```

We have a fork in which we removed `removeAllListeners()` on `this.req` and it fixes the problem.

I tried the subtler option of only removing the listeners that are registered by the module, but removing the module's `error` handler causes the same issue. I tried running various parts of `destroy()`, `cleanup()`, or `preCleanup()` in the next tick but to no avail.

It is unclear to me under which conditions the events could still trigger after `req` has been nullified. I guess an added protection would be to add a `destroyed` flag on the object and check for it in the event handlers.

I can make a PR if that seems sane.

**Comment by darobin** [comment]
Hi, sorry to prod but I was wondering if you'd given thought to the above? We have been running our fork with `removeAllListeners()` suppressed for a while now and it appears not to create problems while simultaneously removing the crash.

**Comment by jcrugzz** [comment]
@darobin since we do null out the object it seems like this should be reasonably safe if its causing issues. Submit a PR and will merge it

**Comment by jcrugzz** [comment]
Fixed in 2.2.0
